File: app/modules/matches/schemas.py
# -*- coding: utf-8 -*-
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any, TYPE_CHECKING
from datetime import datetime
import enum

from app.modules.users.schemas import User

# 比赛状态枚举
class MatchStatus(str, enum.Enum):
    PREPARING = "preparing"
    ONGOING = "ongoing"
    FINISHED = "finished"
    CANCELLED = "cancelled"

# 队员角色枚举
class MemberRole(str, enum.Enum):
    MAIN = "main"
    SUBSTITUTE = "substitute"
    CAPTAIN = "captain"

# --- 基础Schema ---

class ScoreBase(BaseModel):
    points: int
    user_id: int
    team_id: int  # 对应数据库中的match_team_id字段
    event_data: Optional[Dict[str, Any]] = None

class ScoreCreate(ScoreBase):
    pass

class Score(ScoreBase):
    id: int
    match_game_id: int
    standard_score: Optional[float] = None
    recorded_at: datetime
    
    class Config:
        from_attributes = True

# --- 队员相关Schema ---

class TeamMemberBase(BaseModel):
    user_id: int
    role: Optional[str] = "main"

class TeamMemberCreate(TeamMemberBase):
    pass

class TeamMember(TeamMemberBase):
    id: int
    joined_at: datetime
    match_points: int = 0
    
    class Config:
        from_attributes = True

# --- 比赛队伍成员Schema ---

class MatchTeamMembershipBase(BaseModel):
    user_id: int
    role: Optional[str] = "main"

class MatchTeamMembershipCreate(MatchTeamMembershipBase):
    pass

class MatchTeamMembershipSchema(MatchTeamMembershipBase):
    id: int
    match_team_id: int
    joined_at: datetime
    match_points: int = 0
    
    class Config:
        from_attributes = True

# --- 比赛队伍Schema ---

class MatchTeamBase(BaseModel):
    name: str
    color: Optional[str] = None
    external_team_id: Optional[str] = None

class MatchTeamCreate(MatchTeamBase):
    members: Optional[List[TeamMemberCreate]] = None

class MatchTeamUpdate(BaseModel):
    name: Optional[str] = None
    color: Optional[str] = None
    external_team_id: Optional[str] = None

class MatchTeam(MatchTeamBase):
    id: int
    match_id: int
    total_score: int = 0
    games_played: int = 0
    team_rank: Optional[int] = None
    created_at: datetime
    
    class Config:
        from_attributes = True

# 包含比赛信息的队伍详情Schema
class MatchTeamWithMatch(MatchTeam):
    match_name: str
    match_status: MatchStatus
    
    class Config:
        from_attributes = True

# --- 游戏阵容Schema ---

class GameLineupBase(BaseModel):
    user_id: int
    is_starting: bool = True
    substitute_reason: Optional[str] = None

class GameLineupCreate(GameLineupBase):
    pass

class GameLineup(GameLineupBase):
    id: int
    match_game_id: int
    match_team_id: int
    created_at: datetime
    
    class Config:
        from_attributes = True

# --- 赛程Schema ---

class MatchGameBase(BaseModel):
    game_id: int
    game_order: Optional[int] = 1
    structure_type: Optional[str] = None
    structure_details: Optional[Dict[str, Any]] = None
    multiplier: Optional[float] = 1.0

class MatchGameCreate(MatchGameBase):
    pass

class MatchGameUpdate(BaseModel):
    structure_type: Optional[str] = None
    structure_details: Optional[Dict[str, Any]] = None
    is_live: Optional[bool] = None

class MatchGame(MatchGameBase):
    id: int
    match_id: int
    is_live: bool = False
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None
    multiplier: float = 1.0
    total_standard_score: float = 0.0
    average_standard_score: float = 0.0
    created_at: datetime
    score_events: Optional[List["ScoreEvent"]] = None
    
    class Config:
        from_attributes = True

# --- 比赛Schema ---

class MatchBase(BaseModel):
    name: str
    description: Optional[str] = None
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None
    status: Optional[MatchStatus] = MatchStatus.PREPARING
    prize_pool: Optional[str] = None
    max_teams: Optional[int] = None
    max_players_per_team: Optional[int] = 4
    allow_substitutes: Optional[bool] = True
    winning_team_id: Optional[int] = None

class MatchCreate(MatchBase):
    match_games: Optional[List[MatchGameCreate]] = None

class MatchUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None
    status: Optional[MatchStatus] = None
    prize_pool: Optional[str] = None
    max_teams: Optional[int] = None
    max_players_per_team: Optional[int] = None
    allow_substitutes: Optional[bool] = None
    winning_team_id: Optional[int] = None

class Match(MatchBase):
    id: int
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True

class MatchList(MatchBase):
    """Simple match schema for list endpoints without relationships"""
    id: int
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True

# --- 细粒度小分事件 ---

class ScoreEventBase(BaseModel):
    match_game_id: int
    match_team_id: int
    event_type: str
    points: int
    match_id: Optional[int] = None
    opponent_team_id: Optional[int] = None
    user_id: Optional[int] = None
    score_id: Optional[int] = None
    raw_points: Optional[int] = None
    multiplier_used: Optional[float] = None
    tournament_stage: Optional[str] = None
    tournament_round_index: Optional[int] = None
    game_round_label: Optional[str] = None
    game_round_index: Optional[int] = None
    area: Optional[str] = None
    event_time: Optional[datetime] = None
    meta: Optional[Dict[str, Any]] = None

class ScoreEventCreate(ScoreEventBase):
    pass

class ScoreEvent(ScoreEventBase):
    id: int
    created_at: datetime
    match_game: Optional["MatchGame"] = None
    team: Optional[MatchTeam] = None
    opponent_team: Optional[MatchTeam] = None
    user: Optional["User"] = None

    class Config:
        from_attributes = True

class MatchGameEventGroup(BaseModel):
    match_game_id: int
    game_id: int
    game_name: str
    game_code: Optional[str] = None
    events: List[ScoreEvent] = []

class MatchEventsResponse(BaseModel):
    match_id: int
    games: List[MatchGameEventGroup] = []

# 处理前向引用
Score.model_rebuild()
MatchGame.model_rebuild()
ScoreEvent.model_rebuild()

# --- 特殊操作Schema ---

class LineupSetting(BaseModel):
    """设置阵容的Schema"""
    team_lineups: Dict[int, List[int]]  # team_id -> [user_id, user_id, ...]
    substitute_info: Optional[Dict[str, str]] = None  # user_id -> reason

class BatchTeamCreate(BaseModel):
    """批量创建队伍的Schema"""
    teams: List[MatchTeamCreate]

class MemberRoleUpdate(BaseModel):
    """更新队员角色的Schema"""
    role: str  # "main" | "substitute" | "captain"

# --- 兼容性Schema ---

class PlayerMatchStats(BaseModel):
    """玩家在特定比赛中的统计"""
    match_id: int
    match_name: str
    total_points: int
    games_played: int
    team_name: Optional[str] = None

class PlayerStats(BaseModel):
    """玩家总体统计"""
    user_id: int
    nickname: str
    total_matches: int
    total_wins: int
    total_points: int
    win_rate: float
    average_score: float
    current_team: Optional[str] = None
    match_history: List[PlayerMatchStats] = []

# --- 视频相关Schema ---

class VideoPlatform(str, enum.Enum):
    BILIBILI = "bilibili"
    YOUTUBE = "youtube"
    TWITCH = "twitch"
    DOUYU = "douyu"
    HUYA = "huya"
    OTHER = "other"

class VideoType(str, enum.Enum):
    LIVESTREAM = "livestream"
    REPLAY = "replay"
    HIGHLIGHT = "highlight"

class MatchVideoBase(BaseModel):
    title: str
    url: str
    platform: VideoPlatform
    video_type: VideoType = VideoType.REPLAY
    is_official: bool = False
    uploader_name: Optional[str] = None
    match_game_id: Optional[int] = None
    user_id: Optional[int] = None
    description: Optional[str] = None
    duration: Optional[int] = None
    thumbnail_url: Optional[str] = None
    view_count: Optional[int] = None

class MatchVideoCreate(MatchVideoBase):
    pass

class MatchVideoUpdate(BaseModel):
    title: Optional[str] = None
    url: Optional[str] = None
    platform: Optional[VideoPlatform] = None
    video_type: Optional[VideoType] = None
    is_official: Optional[bool] = None
    uploader_name: Optional[str] = None
    match_game_id: Optional[int] = None
    user_id: Optional[int] = None
    description: Optional[str] = None
    duration: Optional[int] = None
    thumbnail_url: Optional[str] = None
    view_count: Optional[int] = None

class MatchVideo(MatchVideoBase):
    id: int
    match_id: int
    view_count: int = 0
    published_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    
    # 关联对象简略信息
    user: Optional[User] = None
    
    class Config:
        from_attributes = True

File: frontend/src/types/schemas.ts
import { z } from 'zod';

// OpenAPI Schema: User
export const UserSchema = z.object({
  id: z.number(),
  nickname: z.string(),
  display_name: z.string().nullable(),
  source: z.string(),
  total_matches: z.number(),
  total_wins: z.number(),
  total_points: z.number(),
  total_standard_score: z.number().optional(),
  average_standard_score: z.number().optional(),
  win_rate: z.number().optional(),
  average_score: z.number().optional(),
  game_level: z.string().optional(),
  level_progress: z.number().optional(),
  created_at: z.string(),
  last_active: z.string(),
});

// Represents MatchTeamMembership
export const MatchTeamMembershipSchema = z.object({
    id: z.number(),
    match_team_id: z.number(),
    user_id: z.number(),
    role: z.string(), // Assuming role is a string from enum
    joined_at: z.string(),
    user: UserSchema,
});

// Represents MatchTeam
export const MatchTeamSchema = z.object({
    id: z.number(),
    match_id: z.number(),
    name: z.string(),
    color: z.string().nullable(),
    total_score: z.number(),
    games_played: z.number(),
    team_rank: z.number().nullable(),
    created_at: z.string(),
    memberships: z.array(MatchTeamMembershipSchema).optional(),
});

// MatchTeam with match information
export const MatchTeamWithMatchSchema = z.object({
    id: z.number(),
    match_id: z.number(),
    name: z.string(),
    color: z.string().nullable(),
    total_score: z.number(),
    games_played: z.number(),
    created_at: z.string(),
    match_name: z.string(),
    match_status: z.string(),
    memberships: z.array(MatchTeamMembershipSchema).optional(),
});

// OpenAPI Schema: Game
export const GameSchema = z.object({
  id: z.number(),
  name: z.string(),
  code: z.string(),
  description: z.string().nullable(),
});

// Represents Score (simplified for API responses)
export const ScoreSchema = z.object({
  id: z.number(),
  points: z.number(),
  user_id: z.number(),
  team_id: z.number(),  // 注意：前端使用team_id，后端model中为match_team_id
  match_game_id: z.number(),
  event_data: z.any().nullable(),
  recorded_at: z.string(),
  user: UserSchema,
  team: MatchTeamSchema,
});

// Represents MatchGame (simplified for API responses)
export const MatchGameSchema = z.object({
  id: z.number(),
  match_id: z.number(),
  game_id: z.number(),
  game_order: z.number(),
  structure_type: z.string().nullable(),
  structure_details: z.record(z.string(), z.any()).nullable(),
  start_time: z.string().nullable(),
  end_time: z.string().nullable(),
  created_at: z.string(),
});

// Simplified Match schema for list endpoints (matches MatchList in backend)
export const MatchListSchema = z.object({
  id: z.number(),
  name: z.string(),
  description: z.string().nullable(),
  start_time: z.string().nullable(),
  end_time: z.string().nullable(),
  status: z.string(), // Assuming status is a string from enum
  prize_pool: z.string().nullable(),
  max_teams: z.number().nullable(),
  max_players_per_team: z.number(),
  allow_substitutes: z.boolean(),
  created_at: z.string(),
  updated_at: z.string(),
});

// The full Match schema for detailed endpoints 
export const MatchSchema = z.object({
  id: z.number(),
  name: z.string(),
  description: z.string().nullable(),
  start_time: z.string().nullable(),
  end_time: z.string().nullable(),
  status: z.string(), // Assuming status is a string from enum
  prize_pool: z.string().nullable(),
  max_teams: z.number().nullable(),
  max_players_per_team: z.number(),
  allow_substitutes: z.boolean(),
  created_at: z.string(),
  updated_at: z.string(),
  winning_team_id: z.number().nullable().optional(),
  is_archived: z.boolean().optional(),
});

// --- API Response Schemas ---
export const MatchesApiResponseSchema = z.array(MatchListSchema);

// --- WebSocket Schemas (from previous knowledge) ---
export const TeamSubScoreSchema = z.object({
  team_id: z.number(),
  team_name: z.string(),
  score: z.number(),
});


// --- Create Schemas for API ---

export const TeamMemberCreateSchema = z.object({
  user_id: z.number(),
  role: z.string().default('main'),
});

export const MatchTeamCreateSchema = z.object({
  name: z.string().min(2),
  color: z.string().optional(),
  members: z.array(TeamMemberCreateSchema).optional(),
});

export const TeamMemberSchema = z.object({
  id: z.number(),
  user_id: z.number(),
  role: z.string(),
  joined_at: z.string(),
  user: UserSchema.optional(),
});

export const GameLineupSchema = z.object({
  id: z.number(),
  match_game_id: z.number(),
  match_team_id: z.number(),
  user_id: z.number(),
  is_starting: z.boolean(),
  substitute_reason: z.string().nullable(),
  created_at: z.string(),
  user: UserSchema.optional(),
});

// User Stats Schema
export const UserStatsSchema = z.object({
  user: z.object({
    id: z.number(),
    nickname: z.string(),
    display_name: z.string().nullable(),
    total_points: z.number(),
    win_rate: z.number(),
    created_at: z.string().nullable(),
    last_active: z.string().nullable(),
  }),
  current_team: z.any().nullable(),
  historical_teams: z.array(z.any()).default([]),
  match_history: z.array(z.object({
    match_id: z.number(),
    match_name: z.string(),
    total_points: z.number(),
    games_played: z.number(),
    team_name: z.string(),
  })).default([]),
  recent_scores: z.array(z.object({
    points: z.number(),
    game_name: z.string(),
    team_name: z.string(),
    recorded_at: z.string(),
  })).default([]),
  // 按站成绩时间线（用于折线图与变化显示）
  score_timeline: z.array(z.object({
    match_id: z.number(),
    match_name: z.string(),
    timestamp: z.string().nullable(),
    avg_standard_score: z.number(),
    rank: z.number().nullable(),
    rank_change: z.number().nullable(),
    score_delta: z.number().nullable(),
  })).default([]),
  game_scores: z.record(z.string(), z.object({
    total_score: z.number(),
    games_played: z.number(),
    game_name: z.string(),  // 添加游戏名称用于显示
    total_standard_score: z.number().optional(),
    average_standard_score: z.number().optional(),
    level: z.string().optional(),
    level_progress: z.number().optional(),
  })).default({}),
});